scenario_player/runner.py

```diff
@@ -153,279 +256,260 @@ def local_seed(self) -> str:
             seed = seed_file.read_text().strip()
         return seed
 
-    def select_chain(self, chain_urls: Dict[str, List[str]]) -> Tuple[str, List[str]]:
-        """Select a chain and return its name and RPC URL.
+    def ensure_token_network_discovery(self) -> ChecksumAddress:
+        """Ensure that all our nodes have discovered the same token network."""
+        discovered = None
 
-        If the currently loaded scenario's designated chain is set to 'any',
-        we randomly select a chain from the given `chain_urls`.
-        Otherwise, we will return `ScenarioRunner.scenario.chain_name` and whatever value
-        may be associated with this key in `chain_urls`.
+        token_network_addresses: Set[ChecksumAddress] = set()
+        for node in self.node_controller:  # type: ignore
+            node_endpoint = API_URL_TOKEN_NETWORK_ADDRESS.format(
+                protocol=self.protocol, target_host=node.base_url, token_address=self.token.address
+            )
+            address = wait_for_token_network_discovery(
+                node_endpoint, self.definition.settings, self.session
+            )
+            token_network_addresses.add(address)
 
-        :raises ScenarioError:
-            if ScenarioRunner.scenario.chain_name is not one of `('any', 'Any', 'ANY')`
-            and it is not a key in `chain_urls`.
-        """
-        chain_name = self.definition.settings.chain
-        if chain_name in ("any", "Any", "ANY"):
-            chain_name = random.choice(list(chain_urls.keys()))
+            log.info("Token Network Discovery", node=node._index, network=discovered)
 
-        log.info("Using chain", chain=chain_name)
-        try:
-            return chain_name, chain_urls[chain_name]
-        except KeyError:
-            raise ScenarioError(
-                f'The scenario requested chain "{chain_name}" for which no RPC-URL is known.'
+        if len(token_network_addresses) != 1:
+            raise RuntimeError(
+                f"Nodes diverged on the token network address, there should be "
+                f"exactly one token network available for all nodes. Current "
+                f"values : {token_network_addresses}"
             )
 
-    def wait_for_token_network_discovery(self, node) -> ChecksumAddress:
-        """Check for token network discovery with the given `node`.
+        return token_network_addresses.pop()
 
-        By default exit the wait if the token has not been discovered after `n` seconds,
-        where `n` is the value of :attr:`.timeout`.
+    def run_scenario(self):
+        self.node_controller.initialize_nodes()
```

Comment:
The node controller initialization used to be spread out across different threads [here](https://github.com/raiden-network/scenario-player/blob/f208ed6cb4f713c1efc921b10458365bcf918b5d/scenario_player/runner.py#L381-L383) and [here](https://github.com/raiden-network/scenario-player/blob/f208ed6cb4f713c1efc921b10458365bcf918b5d/scenario_player/runner.py#L296-L297). This places the initialization in the same thread as the test and groups the calls to the controller.